### PR TITLE
NFT Implementation

### PR DESCRIPTION
**Pull request by Fraccaman**

### CODE REVIEW

**Comment by Fraccaman** [comment]
pls update wasm

File: wasm/wasm_source/Makefile [hunk 1]
```diff
@@ -10,12 +10,14 @@ wasms += tx_init_account
...
 wasms += tx_from_intent
 wasms += tx_transfer
+wasms += tx_create_nft
 wasms += tx_unbond
 wasms += tx_update_vp
 wasms += tx_withdraw
 wasms += vp_testnet_faucet
 wasms += vp_token
 wasms += vp_user
+wasms += vp_nft
```
Comment (tzemanovic) [inline wasm/wasm_source/Makefile:20]: the `tx_mint_nft_tokens` is missing here and in Cargo.toml

File: apps/src/lib/client/tx.rs [hunk 2]
```diff
@@ -420,6 +426,81 @@ pub async fn submit_transfer(ctx: Context, args: args::TxTransfer) {
...
+        opt_keys: nft.opt_keys,
+        tokens: nft.tokens,
+    };
+
+    let data = data.try_to_vec().expect(
+        "Encoding transfer data to initialize a new account shouldn't fail",
+    );
+
+    let tx_code = std::fs::read(TX_CREATE_NFT)
+        .expect("Expected a file at given code path");
```
Comment (tzemanovic) [inline apps/src/lib/client/tx.rs:455]: These have to be read via the wasm loader from ctx:
```suggestion
    let tx_code = ctx.read_wasm(TX_CREATE_NFT);
```

File: apps/src/lib/client/tx.rs [hunk 3]
```diff
@@ -420,6 +426,81 @@ pub async fn submit_transfer(ctx: Context, args: args::TxTransfer) {
...
+        creator: nft_creator_address,
+        tokens: nft_tokens,
+    };
+
+    let data = data.try_to_vec().expect(
+        "Encoding transfer data to initialize a new account shouldn't fail",
+    );
+
+    let tx_code = std::fs::read(TX_MINT_NFT_TOKEN)
+        .expect("Expected a file at given code path");
```
Comment (tzemanovic) [inline apps/src/lib/client/tx.rs:495]: same here
```suggestion
    let tx_code = ctx.read_wasm(TX_MINT_NFT_TOKEN);
```

File: apps/src/lib/cli.rs [hunk 4]
```diff
@@ -987,6 +1004,50 @@ pub mod cmds {
         }
     }
 
+    #[derive(Clone, Debug)]
+    pub struct TxNftCreate(pub args::NftCreate);
+
+    impl SubCmd for TxNftCreate {
+        const CMD: &'static str = "nft-create";
```
Comment (tzemanovic) [inline apps/src/lib/cli.rs:1011]: tiny nit, but just to be consistent with `init-account` and `init-validator`, I'd name this one:
```suggestion
        const CMD: &'static str = "init-nft";
```

File: apps/src/lib/cli.rs [hunk 5]
```diff
@@ -987,6 +1004,50 @@ pub mod cmds {
...
+                .about("Create a new NFT.")
+                .add_args::<args::NftCreate>()
+        }
+    }
+
+    #[derive(Clone, Debug)]
+    pub struct TxNftMint(pub args::NftMint);
+
+    impl SubCmd for TxNftMint {
+        const CMD: &'static str = "nft-mint";
```
Comment (tzemanovic) [inline apps/src/lib/cli.rs:1033]: And then
```suggestion
        const CMD: &'static str = "mint-nft";
```

File: apps/src/lib/client/tx.rs [hunk 6]
```diff
@@ -420,6 +426,81 @@ pub async fn submit_transfer(ctx: Context, args: args::TxTransfer) {
...
+    let nft: Nft = serde_json::from_reader(file)
+        .expect("Couldn't deserialize nft data file");
+
+    let vp_code = match &nft.vp_path {
+        Some(path) => {
+            std::fs::read(path).expect("Expected a file at given code path")
+        }
+        None => ctx.read_wasm(VP_NFT),
+    };
+
```
Comment (tzemanovic) [inline apps/src/lib/client/tx.rs:440]: It's a bit of a hack, but it's a nicer UX - we can default the signer to the NFT creator with:
```suggestion
    let signer = Some(WalletAddress::new( nft.creator.to_string()));
```
(1/2)

File: apps/src/lib/client/tx.rs [hunk 7]
```diff
@@ -420,6 +426,81 @@ pub async fn submit_transfer(ctx: Context, args: args::TxTransfer) {
...
+    let data = data.try_to_vec().expect(
+        "Encoding transfer data to initialize a new account shouldn't fail",
+    );
+
+    let tx_code = std::fs::read(TX_CREATE_NFT)
+        .expect("Expected a file at given code path");
+
+    let tx = Tx::new(tx_code, Some(data));
+
+    let (ctx, tx, keypair) = sign_tx(ctx, tx, &args.tx, None).await;
```
Comment (tzemanovic) [inline apps/src/lib/client/tx.rs:459]: (2/2)
```suggestion
    let (ctx, tx, keypair) = sign_tx(ctx, tx, &args.tx, signer.as_ref()).await;
```

File: apps/src/lib/client/tx.rs [hunk 8]
```diff
@@ -420,6 +426,81 @@ pub async fn submit_transfer(ctx: Context, args: args::TxTransfer) {
...
+    )
+    .await
+    {
+        Some(addr) => addr,
+        None => {
+            eprintln!("No creator key found for {}", &args.nft_address);
+            safe_exit(1);
+        }
+    };
+
```
Comment (tzemanovic) [inline apps/src/lib/client/tx.rs:483]: Similar to nft init - 1/2:
```suggestion
let signer = Some(WalletAddress::new(nft_creator_address.to_string()));
```

File: apps/src/lib/client/tx.rs [hunk 9]
```diff
@@ -420,6 +426,81 @@ pub async fn submit_transfer(ctx: Context, args: args::TxTransfer) {
...
+    let data = data.try_to_vec().expect(
+        "Encoding transfer data to initialize a new account shouldn't fail",
+    );
+
+    let tx_code = std::fs::read(TX_MINT_NFT_TOKEN)
+        .expect("Expected a file at given code path");
+
+    let tx = Tx::new(tx_code, Some(data));
+
+    let (ctx, tx, keypair) = sign_tx(ctx, tx, &args.tx, None).await;
```
Comment (tzemanovic) [inline apps/src/lib/client/tx.rs:499]: and 2/2:
```suggestion
    let (ctx, tx, keypair) = sign_tx(ctx, tx, &args.tx, signer.as_ref()).await;
```

File: apps/src/lib/cli.rs [hunk 10]
```diff
@@ -1669,6 +1731,66 @@ pub mod args {
         }
     }
 
+    // Transaction to create a new nft
+    #[derive(Clone, Debug)]
+    pub struct NftCreate {
+        /// Common tx argumentsips
+        pub tx: Tx,
+        /// Path to the nft file description
+        pub nft_data: PathBuf,
```
Comment (tzemanovic) [inline apps/src/lib/cli.rs:1740]: we'll need some docs about the format of this (later)

File: apps/src/lib/cli.rs [hunk 11]
```diff
@@ -1669,6 +1731,66 @@ pub mod args {
...
+    }
+
+    #[derive(Clone, Debug)]
+    pub struct NftMint {
+        /// Common tx arguments
+        pub tx: Tx,
+        /// The nft address
+        pub nft_address: Address,
+        /// The nft token description
+        pub nft_data: PathBuf,
```
Comment (tzemanovic) [inline apps/src/lib/cli.rs:1767]: and this arg

File: shared/src/types/storage.rs [hunk 12]
```diff
@@ -202,6 +202,21 @@ impl Key {
         Ok(Key { segments })
     }
 
+    /// Returns the last segmenet without popping it
+    pub fn peek(&self) -> Option<&DbKeySeg> {
```
Comment (tzemanovic) [inline shared/src/types/storage.rs:206]: it looks like these are not used, but if we do add them, I think `peek` should be called `last` (same as the method on Vec, which we can also call to implement it) and pop might be better as `pop(&mut self) -> Result<DBKeySeg>`

File: wasm/wasm_source/src/vp_nft.rs [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+
+        let vp_env =
+            init_vp_env_from_tx(nft_address.clone(), tx_env, |address| {
+                // Apply transfer in a transaction
+                tx_host_env::nft::mint_tokens(MintNft {
+                    address: nft_address.clone(),
+                    tokens: vec![],
+                    creator: nft_creator.clone(),
+                });
+                tx_host_env::insert_verifier(address)
```
Comment (tzemanovic) [inline wasm/wasm_source/src/vp_nft.rs:101]: (it's applied inside `mint_tokens`)
```suggestion
```

Comment (tzemanovic) [reply wasm/wasm_source/src/vp_nft.rs:101]: nvm sry (mint tokens only adds the creator's address)

File: wasm/wasm_source/src/vp_nft.rs [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+        let keys_changed: HashSet<storage::Key> =
+            vp_env.all_touched_storage_keys();
+        let verifiers: HashSet<Address> = vp_env.get_verifiers();
+
+        assert!(validate_tx(tx_data, nft_address, keys_changed, verifiers));
+    }
+
+    /// Test that only owner can mint new tokens
+    #[test]
+    fn test_mint_tokens_wrong_owner() {
```
Comment (tzemanovic) [inline wasm/wasm_source/src/vp_nft.rs:165]: there seems to be some non-determinism in the VP with this test. I'm running with `RUST_LOG=debug RUST_BACKTRACE=1 cargo test vp_nft::tests::test_mint_tokens_wrong_owner --features vp_nft -- --nocapture` and occasionally it fails with:
```
thread 'vp_nft::tests::test_mint_tokens_wrong_owner' panicked at 'called `Option::unwrap()` on a `None` value', anoma/vm_env/src/nft.rs:157:41
```

Comment (Fraccaman) [reply wasm/wasm_source/src/vp_nft.rs:165]: ah, didnt notice it! nice catch, thanks :)

File: shared/src/types/storage.rs (hunk 12 above)
Comment (Fraccaman) [reply shared/src/types/storage.rs]: Ill remove them

**Comment by Fraccaman** [comment]
pls update wasm

**Pull request merged**